times_reader/main.py

```diff
@@ -265,20 +269,22 @@ def produce_times_tables(
         else:
             used_tables.add(mapping.xl_name)
             df = input[mapping.xl_name].copy()
-            if mapping.filter_rows:
-                # Select just the rows where the attribute value matches the last mapping column or output table name
-                if not "Attribute" in df.columns:
+            # Filter rows according to filter_rows mapping:
+            for filter_col, filter_val in mapping.filter_rows.items():
+                if filter_col not in df.columns:
                     print(
                         f"WARNING: Cannot produce table {mapping.times_name} because input"
-                        f" table {mapping.xl_name} does not contain an Attribute column"
+                        f" table {mapping.xl_name} does not contain column {filter_col}"
                     )
-                else:
-                    colname = mapping.xl_cols[-1]
-                    filter = set(x.lower() for x in {colname, mapping.times_name})
-                    i = df["Attribute"].str.lower().isin(filter)
-                    df = df.loc[i, :]
-                    if colname not in df.columns:
-                        df = df.rename(columns={"VALUE": colname})
+                    # TODO break this loop and continue outer loop?
+                filter = set(x.lower() for x in {filter_val})
+                i = df[filter_col].str.lower().isin(filter)
+                df = df.loc[i, :]
+                # Only for Attributes, if the attribute name is not a column in the table,
+                # rename VALUE to the attribute name
+                # TODO generalize this and have mapping.rename_cols dict?
+                if filter_col == "Attribute" and filter_val not in df.columns:
```

Comment:
Should we remove special treatment of attributes? We would then need to add `Value` column before `Attribute: ...` in the mapping file where relevant...

Reply:
Yes, nice idea, let me try that. (Also, CI fails because for some reason we have some type errors which didn't show up in my PR introducing the pyright type checker, but let me see if we can fix the type errors -- and improve our code quality.)